cpp/include/raft/util/cudart_utils.hpp

```diff
@@ -537,6 +538,83 @@ inline auto get_pool_memory_resource(rmm::mr::device_memory_resource*& mr, size_
   return pool_res;
 }
 
+/**
+ * RAII helper to access the host data from GPU.
+ * If a valid host pointer is passed to the constructor of this struct, it registers
+ * and maps the host memory region to be accessible from the current device for the
+ * lifetime of this struct.
+ *
+ * This structure is especially useful when the host memory is not allocated within the logic
+ * of the algorithm, e.g. the pointer comes from the OS `mmap` call. In such cases, it helps to
+ * avoid an extra host<->host memory copy.
+ *
+ * Note, the host memory in question must fit within the GPU memory
+ * (the mapping logic cannot produce a UVM/managed memory).
+ */
+template <typename PtrT>
+struct using_mapped_memory_t {
+  static_assert(std::is_pointer_v<PtrT>, "PtrT must be a pointer type");
+
+  /**
+   * When necessary, register the input pointer and make it accessible from GPU.
+   * If the input pointer is a `nullptr` or already accessible from a current device,
+   * the struct does nothing to it.
+   *
+   * Note, when the memory needs to be mapped and the `size` is larger than the available device
+   * memory, the call fails with a raft exception due to `cudaErrorMemoryAllocation` error.
+   *
+   * @param ptr host or device pointer.
+   * @param size amount of the data to be mapped in bytes.
+   */
+  using_mapped_memory_t(PtrT ptr, size_t size)
+  {
+    if (ptr == nullptr) { return; }
+    cudaPointerAttributes attr;
+    RAFT_CUDA_TRY(cudaPointerGetAttributes(&attr, ptr));
+    if (attr.devicePointer != nullptr) {
+      dev_ptr_ = attr.devicePointer;
+    } else {
+      host_ptr_ = reinterpret_cast<void*>(const_cast<uncv_ptr_t>(ptr));
+      RAFT_CUDA_TRY(cudaHostRegister(host_ptr_, size, choose_flags()));
```

Comment:
1. Since using mapped memory has a perf impact, it would be good to emit a log message.
2. It would be great if the error message could inform the user about possible workaround (e.g. using managed memory).

Reply:
I would prefer option 1, maybe a new file in `raft/util/`?

Reply:
Ultimately I'd like to see a builder in raft/core for this if it's part of the vocabulary that users should be interacting with directly.

However, I'd also like to see it consolidated with mdarray. For example, I think it would plug into the existing mdarray API nicely as a new container strategy  and I believe it would be able to make use of the host_device_accessor to represent the ability to be read from both host and device.